### PR TITLE
CustomInterpolator

### PR DESCRIPTION
**Pull request by apoorvagnihotri**
- Added `usage3.py` for example usage of the new interpolation method `CustomInterpolator`.
- Reshaping into `shape(self.resolution, self.resolution)` the output of predict inside `base.py`

### CODE REVIEW

**Comment by sdeepaknarayanan** [comment]
Then do we use this class for Nearest Neighbors too?

**Comment by apoorvagnihotri** [comment]
Yes, I tried, it works! Added to the `usage3.py` for your ref.

**Comment by sdeepaknarayanan** [comment]
What about GP regressor? That's the only other thing dependent on ScikitLearn I suppose?

**Comment by apoorvagnihotri** [comment]
> Also, why are you reshaping twice? Once in return of predict_grid, and once in usage3.py

That would not lead to any incorrectness, in this case; unfortunately, I forgot to remove the reshape in `usage3.py`

**Comment by apoorvagnihotri** [comment]
> What about GP regressor? That's the only other thing dependent on ScikitLearn I suppose?

GP can be removed as a special method. We can just use custom interpolator in that case too.

**Comment by sdeepaknarayanan** [comment]
I think if you can remove GP and Nearest neighbors, we can merge!

**Comment by sdeepaknarayanan** [comment]
> > Also, why are you reshaping twice? Once in return of predict_grid, and once in usage3.py
> 
> That would not lead to any incorrectness, in this case; unfortunately, I forgot to remove the reshape in `usage3.py`

Actually I think there is an interpretation issue with the gridding method. If you run usage.py, you'll see that the interpolation is performed in a OpenCV reading an image like manner. (The X axis grows from left to right and the Y Axis from top to bottom). The point (3,3) is on the bottom right, instead of the natural top right in the case of cartesan coordinates. While the working of the code is perfectly fine and the values are prety close, the issue arises when we try dealing with interpretation

**Comment by apoorvagnihotri** [comment]
> > > Also, why are you reshaping twice? Once in return of predict_grid, and once in usage3.py
> > 
> > 
> > That would not lead to any incorrectness, in this case; unfortunately, I forgot to remove the reshape in `usage3.py`
> 
> Actually I think there is an interpretation issue with the gridding method. If you run usage.py, you'll see that the interpolation is performed in a OpenCV reading an image like manner. (The X axis grows from left to right and the Y Axis from top to bottom). The point (3,3) is on the bottom right, instead of the natural top right in the case of cartesan coordinates. While the working of the code is perfectly fine and the values are prety close, the issue arises when we try dealing with interpretation

That is an extremely good catch! What do you suggest moving on?

**Comment by sdeepaknarayanan** [comment]
For now, I would say lets not worry about it. We can look at it later. I don't think the issue is really major.  I'll merge this PR once you can remove nearest neighbors and gp. I think KNN has another parameter (the weighting function - add that to usage3.py - that'd be very helpful)

**Comment by apoorvagnihotri** [comment]
@sdeepaknarayanan I noticed that [sklearn's KNN](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsRegressor.html#sklearn.neighbors.KNeighborsRegressor) also takes in weights as `'distance'` is it similar to `Idw`?

++ updated the `usage3.py` with weights hyper-parameter.

**Comment by sdeepaknarayanan** [comment]
@apoorvagnihotri Similar but not same. K - defines the number of neighbors. IDW takes all the data points in the training data. IDW is more general (we define the exponent for the distance metric). AFAIK Sklearn uses just the L2 norm and no exponent. While we can make KNN a general part of IDW, I think there will be implementation issues (efficacy amongst others). Better use Sklearn's imho.
